### PR TITLE
[Giuseppe Picciuca] adaptors_task

### PR DESCRIPTION
**Pull request by gpicciuca**
All tests pass.

### CODE REVIEW

File: adaptors_task/student/src/impl.cpp [hunk 1]
```diff
@@ -1,8 +1,50 @@
...
+            stack.push(c);
+            continue;
+        }
+
+        if (it->first == c)
+        {
+            // Opening bracket
+            stack.push(c);
+        }
+        else if (it->second == c)
```
Comment (smay1613) [inline adaptors_task/student/src/impl.cpp]: so please name this condition instead of using comments

File: adaptors_task/student/src/impl.cpp [hunk 2]
```diff
@@ -1,8 +1,50 @@
...
+        {
+            // Closing bracket
+            char lastChar = stack.top();
+
+            // Check if last opening bracket matches closing bracket
+            if (lastChar == it->first)
```
Comment (smay1613) [inline adaptors_task/student/src/impl.cpp]: Usually you should avoid comments in your code - everything should be shown in code, so please, just give a name to your conditions, so there will be no need for a comments

Comment (smay1613) [inline adaptors_task/student/src/impl.cpp]: we can stop the algorithm in another case

File: adaptors_task/student/src/impl.cpp [hunk 3]
```diff
@@ -1,8 +1,50 @@
...
+    for (auto& c : source)
+    {
+        auto it = std::find_if(brackets.begin(), brackets.end(), [&](const std::pair<char, char>& p) -> bool {
+            return p.first == c || p.second == c;
+        });
+
+        if (it == brackets.end())
+            continue; // Not a bracket
+
+        if (stack.empty())
```
Comment (smay1613) [inline adaptors_task/student/src/impl.cpp]: why should you do this?

File: adaptors_task/student/src/impl.cpp [hunk 4]
```diff
@@ -1,8 +1,50 @@
...
     {'{', '}'}
 };
+
+bool isValid(const std::string& source)
+{
+    std::stack<char> stack;
+
+    for (auto& c : source)
+    {
+        auto it = std::find_if(brackets.begin(), brackets.end(), [&](const std::pair<char, char>& p) -> bool {
```
Comment (smay1613) [inline adaptors_task/student/src/impl.cpp]: please try to decompose this function - create helper function like isBracket(),

**Comment by gpicciuca** [comment]
Cleaned up the code and applied your suggestions.

File: adaptors_task/student/src/impl.cpp [hunk 5]
```diff
@@ -1,48 +1,58 @@
...
     {'(', ')'},
     {'[', ']'},
     {'{', '}'}
 };
 
+bool isBracket(char c, char& matchOpenBracket, bool& isOpening, bool& isClosing)
+{
+    for (auto& bracket : brackets)
+    {
+        if (bracket.first != c && bracket.second != c)
```
Comment (smay1613) [inline adaptors_task/student/src/impl.cpp:15]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es77-minimize-the-use-of-break-and-continue-in-loops

File: adaptors_task/student/src/impl.cpp [hunk 6]
```diff
@@ -1,48 +1,58 @@
...
+    return false;
+}
+
 bool isValid(const std::string& source)
 {
     std::stack<char> stack;
 
+    bool isOpeningBracket = false;
+    bool isClosingBracket = false;
+    char matchOpenBracket = '\0';
```
Comment (smay1613) [inline adaptors_task/student/src/impl.cpp:34]: {}

File: adaptors_task/student/src/impl.cpp [hunk 7]
```diff
@@ -1,48 +1,58 @@
...
 #include <stack>
-#include <algorithm>
 
 const std::vector<std::pair<char, char>> brackets {
     {'(', ')'},
     {'[', ']'},
     {'{', '}'}
 };
 
+bool isBracket(char c, char& matchOpenBracket, bool& isOpening, bool& isClosing)
```
Comment (smay1613) [inline adaptors_task/student/src/impl.cpp:11]: maybe we should separate it by 3 different functions

**Pull request closed without merging**